### ISSUE TITLE
Assertion `ptr->_u.st == future_state_base::state::future during compaction on a start of another node

### ISSUE THREAD

**Issue by enaydanov** [scylladb/scylla#10179]
## Installation details

Kernel Version: 5.11.0-1028-aws

Scylla version (or git commit hash): `5.1.dev-20220303.3b5ba5c1a998` with build-id `abc821aa82a471e4414ae67ec146cbffc9c8f194`

Cluster size: 4 nodes (i3en.2xlarge)

Scylla running with shards number (live nodes):
**No resources left at the end of the run**

OS / Image: `ami-04e93503987e389d7` (aws: eu-west-1)


Test: `longevity-twcs-48h-test`

Test id: `42f0a3aa-66c6-4358-9f87-0111cbdbeced`

Test name: `longevity_twcs_test/longevity-twcs-48h-test`

Test config file(s):

- [longevity-twcs-48h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/e4c0bed30cedac33ef36d5d1e58e193669c3d2f5/test-cases/longevity/longevity-twcs-48h.yaml)


## Issue description

Node `longevity-twcs-48h-master-db-node-42f0a3aa-3` got assertion `ptr->_u.st == future_state_base::state::future` during restart of node `longevity-twcs-48h-master-db-node-42f0a3aa-4`:

```
...
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |  [shard 2] reader_concurrency_semaphore - (rate limiting dropped 60 similar messages) Semaphore _read_concurrency_sem with 16/100 count and 13116425/154811760 memory resources: timed out, dumping permit diagnostics:
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO | scylla: ./seastar/include/seastar/core/future.hh:916: void seastar::internal::promise_base_with_type<seastar::rpc::tuple<seastar::foreign_ptr<seastar::lw_shared_ptr<query::result>>, cache_temperature>>::set_urgent_state(seastar::internal::promise_base_with_type::future_state &&) [T = seastar::rpc::tuple<seastar::foreign_ptr<seastar::lw_shared_ptr<query::result>>, cache_temperature>]: Assertion `ptr->_u.st == future_state_base::state::future' failed.
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO | Aborting on shard 2.                                                                                                                            
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO | Backtrace:                                                                                                                                                                                                                                                                                                                   
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45cddc8                                                                                                                                                                                                                                                                                                                 
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45fe952                                                                                                                                                                                                                                                                                                                   
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x7fa90f05ea1f                                                                                                                                                                                                                                                                                                                                                                                                                                         
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x3d2a1                                                                                                                                                                                                                                                                                  
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x268a3                                                                                                                                                                                                                                                                                                    
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x26788                                                                                                                                                                                                                                                                                    
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x35a15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x256993b                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45deda4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45e0187                                                                                                                                                                                                                                                                                                                  
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45fee55                                                                                                                                                                                                                                                                                                                                                  
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   0x45b2caa                                                                                                                                                                                                                                                                                    
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libpthread.so.0+0x92a4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
2022-03-04T08:38:35+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-3 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x100322                                                                                                                                                                                                                                                             
...
```

At the same time on the starting node:

```
...
2022-03-04T08:38:33+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-4 !    INFO |  [shard 6] compaction - [Compact scylla_bench.test 7a2900f0-9b96-11ec-91aa-617c1ca2dcb4] Compacting [/var/lib/scylla/data/scylla_bench/test-e0ba9a809b7411ecbcff687625ba7c88/me-22382-big-Data.db:level=0:origin=compaction,/var/lib/scylla/data/scylla_bench/test-e0ba9a809b7411ecbcff687625ba7c88/me-22654-big-Data.db:level=0:origin=compaction,/var/lib/scylla/data/scylla_bench/test-e0ba9a809b7411ecbcff687625ba7c88/me-27438-big-Data.db:level=0:origin=reshape,/var/lib/scylla/data/scylla_bench/test-e0ba9a809b7411ecbcff687625ba7c88/me-18022-big-Data.db
2022-03-04T08:38:34+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-4 !    INFO |  [shard 2] reader_concurrency_semaphore - Semaphore _read_concurrency_sem with 7/100 count and 23734441/154811760 memory resources: timed out, dumping permit diagnostics:                                                                                                                                                                                                                                                                                                                                                                                         
2022-03-04T08:38:34+00:00 longevity-twcs-48h-master-db-node-42f0a3aa-4 !    INFO |  [shard 6] compaction - [Compact scylla_bench.test 7a2900f0-9b96-11ec-91aa-617c1ca2dcb4] Compacted 4 sstables to [/var/lib/scylla/data/scylla_bench/test-e0ba9a809b7411ecbcff687625ba7c88/me-27758-big-Data.db:level=0]. 5MB to 2MB (~47% of original) in 1128ms = 2MB/s. ~512 total partitions merged to 40.                                                                                                                                                                                                                                                      
...
```

Core file:

gs://upload.scylladb.com/core.scylla.113.92465a04c0bc4a0ebc800efa98c2aadc.9820.1646383115000000000000/core.scylla.113.92465a04c0bc4a0ebc800efa98c2aadc.9820.1646383115000000000000.gz


- Restore Monitor Stack command: `$ hydra investigate show-monitor 42f0a3aa-66c6-4358-9f87-0111cbdbeced`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=42f0a3aa-66c6-4358-9f87-0111cbdbeced)
- Show all stored logs command: `$ hydra investigate show-logs 42f0a3aa-66c6-4358-9f87-0111cbdbeced`



## Logs:
- **db-cluster-42f0a3aa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/db-cluster-42f0a3aa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/db-cluster-42f0a3aa.tar.gz)
- **monitor-set-42f0a3aa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/monitor-set-42f0a3aa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/monitor-set-42f0a3aa.tar.gz)
- **loader-set-42f0a3aa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/loader-set-42f0a3aa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/loader-set-42f0a3aa.tar.gz)
- **sct-runner-42f0a3aa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/sct-runner-42f0a3aa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/42f0a3aa-66c6-4358-9f87-0111cbdbeced/20220304_090518/sct-runner-42f0a3aa.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-twcs-48h-test/22/)

**Comment by enaydanov** [comment]
Monitoring node still alive: http://54.194.204.47:3000/

**Comment by denesb** [comment]
@slivne the crash happens because someone is trying to wait on an already waited-on future. There is no future waiting in the diagnostics dump. I already noted in another issue that sct is swallowing the meat of the diagnostics dump. This used to work, so this was broken recently.
I will have a look at the core.

**Comment by bhalevy** [comment]
Decoded backtrace:
```
void seastar::backtrace(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:750
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:780
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:792
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3668
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3650
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3646
?? ??:0
/data/scylla-s3-reloc.cache/by-build-id/abc821aa82a471e4414ae67ec146cbffc9c8f194/extracted/scylla/libexec/scylladb/libreloc/libc.so.6 0x3d2a1 
/data/scylla-s3-reloc.cache/by-build-id/abc821aa82a471e4414ae67ec146cbffc9c8f194/extracted/scylla/libexec/scylladb/libreloc/libc.so.6 0x268a3 
/data/scylla-s3-reloc.cache/by-build-id/abc821aa82a471e4414ae67ec146cbffc9c8f194/extracted/scylla/libexec/scylladb/libreloc/libc.so.6 0x26788 
/data/scylla-s3-reloc.cache/by-build-id/abc821aa82a471e4414ae67ec146cbffc9c8f194/extracted/scylla/libexec/scylladb/libreloc/libc.so.6 0x35a15 
seastar::internal::promise_base_with_type >, cache_temperature> >::set_urgent_state(seastar::future_state >, cache_temperature> >&&) at ././seastar/include/seastar/core/future.hh:916
 (inlined by) seastar::future >, cache_temperature> >::forward_to(seastar::internal::promise_base_with_type >, cache_temperature> >&&) at ././seastar/include/seastar/core/future.hh:1696
 (inlined by) void seastar::futurize >, cache_temperature> > >::satisfy_with_result_of >, cache_temperature> >::then_wrapped_nrvo >, cache_temperature> >, seastar::future >, cache_temperature> >::finally_body::invoke_on, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31, , seastar::future >, cache_temperature> > >(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}, false> >(seastar::future >, cache_temperature> >::finally_body::invoke_on, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31, , seastar::future >, cache_temperature> > >(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}, false>&&)::{lambda(seastar::internal::promise_base_with_type >, cache_temperature> >&&, {lambda()#1}&, seastar::future_state >, cache_temperature> >&&)#1}::operator()(seastar::internal::promise_base_with_type, seastar::internal::promise_base_with_type >, cache_temperature> >, seastar::future_state) const::{lambda()#1}>(seastar::internal::promise_base_with_type, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&) at ././seastar/include/seastar/core/future.hh:2120
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1667
 (inlined by) seastar::continuation >, cache_temperature> >, seastar::future >, cache_temperature> >::finally_body::invoke_on, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31, , seastar::future >, cache_temperature> > >(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, service::storage_proxy::query_result_local(seastar::lw_shared_ptr, seastar::lw_shared_ptr, nonwrapping_interval const&, query::result_options, tracing::trace_state_ptr, std::chrono::time_point > >)::$_31&&)::{lambda()#1}, false>, seastar::future >, cache_temperature> >::then_wrapped_nrvo >, cache_temperature> >, {lambda()#1}>({lambda()#1}&&)::{lambda(seastar::internal::promise_base_with_type >, cache_temperature> >&&, {lambda()#1}&, seastar::future_state >, cache_temperature> >&&)#1}, seastar::rpc::tuple >, cache_temperature> >::run_and_dispose() at ././seastar/include/seastar/core/future.hh:767
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2344
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2754
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2923
operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:4142
 (inlined by) void std::__invoke_impl(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) std::enable_if, void>::type std::__invoke_r(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_89&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:111
 (inlined by) std::_Function_handler::_M_invoke(std::_Any_data const&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:290
std::function::operator()() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:590
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:60
```

**Comment by avikivity** [comment]
Why do the function names look so strange in the trace?

**Comment by denesb** [comment]
The crash seems to happen here:
https://github.com/scylladb/seastar/blob/4e42a60199a26b3e784459c0ba476852e5deb371/include/seastar/core/smp.hh#L350
```c++
                    // Deferring call on rvalue function, make sure to preserve it across call
                    auto w = std::make_unique<std::decay_t<Func>>(std::move(func));
                    auto ret = futurize<ret_type>::invoke(*w);
                    return ret.finally([w = std::move(w)] {});
```
I have no idea why yet. This is seastar code, I have not yet seen such errors in seastar code.

**Comment by denesb** [comment]
In the context of https://github.com/scylladb/scylla/blob/f2060ac03bf008b40bc8d0ece005b79745ca593b/service/storage_proxy.cc#L4045-L4071

**Comment by denesb** [comment]
This looks like it could be related to the recent exception propagating rework of storage proxy, but no hard proof yet.
CC @piodul

**Comment by piodul** [comment]
@denesb `storage_proxy::query_result_local` is replica-side logic. I didn't do any changes there related to exception propagation, so far I only touched the coordinator part.

**Comment by denesb** [comment]
In the core, the coordinator==replica, so all this happens on the same node and furthermore the same shard.

**Comment by bhalevy** [comment]
> Why do the function names look so strange in the trace?

I copied the decoded backtrace from backtrace.scylladb.com
The server uses preformatted output for the backtrace part so maybe something in the names screws up the html output.

**Comment by avikivity** [comment]
It could be this assertion (but not seen in logs?)

```c++
    void set_urgent_state(future_state&& state) noexcept {
        auto* ptr = get_state();
        // The state can be null if the corresponding future has been
        // destroyed without producing a continuation.
        if (ptr) {
            // FIXME: This is a fairly expensive assert. It would be a
            // good candidate for being disabled in release builds if
            // we had such an assert.
            assert(ptr->_u.st == future_state_base::state::future);
            new (ptr) future_state(std::move(state));
            make_ready<urgent::yes>();
        }
    }
```

if so, it is likely memory corruption.

**Comment by bhalevy** [comment]
> > Why do the function names look so strange in the trace?
> 
> I copied the decoded backtrace from backtrace.scylladb.com The server uses preformatted output for the backtrace part so maybe something in the names screws up the html output.

@avikivity indeed the pre-formatted output of the decoded backtrace includes `[&<>]` characters therefore it need html escaping.
Fixed in https://github.com/scylladb/scylla_s3_reloc_server/pull/10
and already applied on the server .

**Comment by denesb** [comment]
> It could be this assertion (but not seen in logs?)
> 

According to gdb it is.

> 
> if so, it is likely memory corruption.

Yes, very likely, although all pointers I found so far seems to be valid (which doesn't exclude memory corruption).

**Comment by avikivity** [comment]
What is ptr->_u.st? Apart from "optimized out".

**Comment by denesb** [comment]
Well it's complicated (of course):
```
(gdb) p ptr
$44 = (seastar::future_state<seastar::rpc::tuple<seastar::foreign_ptr<seastar::lw_shared_ptr<query::result> >, cache_temperature> > *) 0x0
(gdb) p *this
$45 = {
  <seastar::internal::promise_base> = {
    _future = 0x0,
    _state = 0x60100b4bde70,
    _task = 0x60100b4bde60
  }, 
  members of seastar::internal::promise_base_with_type<seastar::rpc::tuple<seastar::foreign_ptr<seastar::lw_shared_ptr<query::result> >, cache_temperature> >:
  static copy_noexcept = <optimized out>
}
(gdb) p *this->_state
$46 = {
  _u = {
    st = seastar::future_state_base::state::result,
    ex = {
      _M_exception_object = 0x3
    }
  }
}
```
`ptr` being null contradicts the execution path -- we are in the body of an if, entered only if it is not null. Interestingly, `_state`, which is actually what `ptr` is (+ a cast) is not null.

**Comment by avikivity** [comment]
I don't think we can make any progress. This is well established code that has run for trillions of times. The real cause is likely far removed. Let's see if it ever reproduces.

**Comment by denesb** [comment]
I concur. I tried to dig around to see if anything smells, but I failed to get any promising leads.
The `future_state` object `_state` points to is in the next `task` object in the chain, so we can pretty much rule out use-after-free (also supported by `scylla ptr`).

**Comment by denesb** [comment]
> I concur. I tried to dig around to see if anything smells, but I failed to get any promising leads. The `future_state` object `_state` points to is in the next `task` object in the chain, so we can pretty much rule out use-after-free (also supported by `scylla ptr`).

The code causing the memory corruption on the other hand can be in a completely unrelated code-path.

**Comment by bhalevy** [comment]
Closing as duplicate of #10193 
Please reopen if seen again
